app/controllers/passwords_controller.rb

```diff
@@ -78,22 +65,44 @@ def ensure_email_present
   end
 
   def validate_confirmation_token
-    @user = User.find_by(confirmation_token: params[:token].to_s)
-    redirect_to root_path, alert: t("passwords.edit.token_failure") unless @user&.valid_confirmation_token?
+    confirmation_token = params.permit(:token).fetch(:token, "").to_s
+    @user = User.find_by(confirmation_token:)
+    return login_failure(t("passwords.edit.token_failure")) unless @user&.valid_confirmation_token?
+    sign_out if signed_in? && @user != current_user
   end
 
-  def login_failure(message)
-    flash.now.alert = message
-    render template: "multifactor_auths/prompt", status: :unauthorized
+  def password_reset_session_verified
+    reset_session
+    session[:password_reset_verified_user] = @user.id
+    session[:password_reset_verified] = Gemcutter::PASSWORD_VERIFICATION_EXPIRY.from_now
+    @user.update!(confirmation_token: nil)
```

Comment:
should we do this first, in case there are any exceptions modifying the session?

Reply:
I'm going to go with the following order, with the justification that I'd rather leave the app in the most secure state (reset session, no verified session) if for some reason updating the user to invalidate the token fails.

```ruby
  def password_reset_session_verified
    reset_session
    @user.update!(confirmation_token: nil)
    session[:password_reset_verified_user] = @user.id
    session[:password_reset_verified] = Gemcutter::PASSWORD_VERIFICATION_EXPIRY.from_now
  end
```